Microsoft.Toolkit.Uwp.UI/Extensions/Tree/LogicalTree.cs

```diff
@@ -18,294 +20,669 @@ namespace Microsoft.Toolkit.Uwp.UI.Extensions
+        /// Find the first child element matching a given predicate, using a depth-first search.
+        /// </summary>
+        /// <typeparam name="T">The type of elements to match.</typeparam>
+        /// <typeparam name="TPredicate">The type of predicate in use.</typeparam>
+        /// <param name="element">The root element.</param>
+        /// <param name="predicate">The predicatee to use to match the child nodes.</param>
+        /// <returns>The child that was found, or <see langword="null"/>.</returns>
+        private static T? FindChild<T, TPredicate>(this FrameworkElement element, ref TPredicate predicate)
+            where T : notnull, FrameworkElement
+            where TPredicate : struct, IPredicate<T>
+        {
+            // Jump label to manually optimize the tail recursive paths for elements with a single
+            // child by just overwriting the current element and jumping back to the start of the
+            // method. This avoids a recursive call and one stack frame every time.
+            Start:
+
+            if (element is Panel panel)
             {
-                foreach (var child in (element as Panel).Children)
+                foreach (UIElement child in panel.Children)
                 {
-                    var result = (child as FrameworkElement)?.FindChildByName(name);
-                    if (result != null)
+                    if (child is not FrameworkElement current)
+                    {
+                        continue;
+                    }
+
+                    if (child is T result && predicate.Match(result))
                     {
                         return result;
                     }
+
+                    T? descendant = FindChild<T, TPredicate>(current, ref predicate);
+
+                    if (descendant is not null)
+                    {
+                        return descendant;
+                    }
                 }
             }
-            else if (element is ItemsControl)
+            else if (element is ItemsControl itemsControl)
             {
-                foreach (var item in (element as ItemsControl).Items)
+                foreach (object item in itemsControl.Items)
                 {
-                    var result = (item as FrameworkElement)?.FindChildByName(name);
-                    if (result != null)
+                    if (item is not FrameworkElement current)
+                    {
+                        continue;
+                    }
+
+                    if (item is T result && predicate.Match(result))
                     {
                         return result;
                     }
+
+                    T? descendant = FindChild<T, TPredicate>(current, ref predicate);
+
+                    if (descendant is not null)
+                    {
+                        return descendant;
+                    }
                 }
             }
-            else
+            else if (element is UserControl userControl)
```

Comment:
I think we put UserControl to the last branch before the reflection case and make the order: Panel, ItemsControl, ContentControl, Border, UserControl, then reflection.

Here and in FindChildren of course. :)